text/0000-cargo-prepublish.md

```diff
@@ -0,0 +1,395 @@
+
+- Crate `foo` lists dependency `xml-rs = "0.9.0"`
+- Crate `bar` lists dependency `xml-rs = "0.9.1"`
+- Crate `baz` lists dependency `xml-rs = "0.8.0"`
+- Crate `servo` has `foo`, `bar` and `baz` as dependencies.
+
+With this setup, the dependency graph for Servo will contain *two* versions of
+`xml-rs`: `0.9.1` and `0.8.0`. That's because minor versions are coalesced;
+`0.9.1` is considered a minor release against `0.9.0`, while `0.9.0` and `0.8.0`
+are incompatible.
+
+### Scenario: augmenting with a bugfix
+
+Let's say that while developing `foo` we've got a lock file pointing to `xml-rs`
+`0.8.0`, and there's an `0.8.0` branch of `xml-rs` that hasn't been touched
+since it was published. We then find a bug in the 0.8.0 publication of `xml-rs`
+which we'd like to fix.
+
+First we'll check out `foo` locally and implement what we believe is a fix for
+this bug, and next, we change `Cargo.toml` for `foo`:
+
+```toml
+[augment.crates-io]
+xml-rs = { path = "../xml-rs" }
+```
+
+When compiling `foo`, Cargo will resolve the `xml-rs` dependency to `0.8.0`,
+as it did before, but that version's been replaced with our local copy. The
+local path dependency, which has version 0.8.0, takes precedence over the
+version found in the registry.
+
+Once we've confirmed a fix bug we then continue to run tests in `xml-rs` itself,
+and then we'll send a PR to the main `xml-rs` repo. This then leads us to the
+next section where a new version of `xml-rs` comes into play!
+
+### Scenario: prepublishing a new minor version
+
+Now, suppose that `foo` needs some changes to `xml-rs`, but we want to check
+that all of Servo compiles before pushing the changes through.
+
+First, we change `Cargo.toml` for `foo`:
+
+```toml
+[augment.crates-io]
+xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.9.2" }
+
+[dependencies]
+xml-rs = "0.9.2"
+```
+
+For `servo`, we also need to record the prepublication, but don't need to modify
+or introduce any `xml-rs` dependencies; it's enough to be using the fork of
+`foo`, which we would be anyway:
+
+```toml
+[augment.crates-io]
+xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.9.2" }
+
+[dependencies]
+foo = { git = "https://github.com/aturon/foo", branch = "fix-xml" }
+```
+
+With this setup:
+
+- When compiling `foo`, Cargo will resolve the `xml-rs` dependency to `0.9.2`, and
+retrieve the source from the specified git branch.
+
+- When compiling `servo`, Cargo will again resolve *two* versions of `xml-rs`,
+this time `0.9.2` and `0.8.0`, and for the former it will use the source from
+the git branch.
+
+The Cargo.toml files that needed to be changed here span from the crate that
+actually cares about the new version (`foo`) upward to the root of the crate we
+want to do integration testing for (`servo`); no sibling crates needed to be
+changed.
+
+Once `xml-rs` version `0.9.2` is actually published, we can remove the
+`[augment]` sections, and Cargo will warn us that this needs to be done.
```

Comment:
Ah yes sorry I forgot to update that when updating the RFC, I'll correct this to account that no warnings will be emitted and that this may or may not be all that needs to be done.